oap-shuffle/remote-shuffle/src/main/scala/org/apache/spark/shuffle/sort/remote/HadoopFileSegmentManagedBuffer.scala

```
@@ -0,0 +1,207 @@
+
+  import HadoopFileSegmentManagedBuffer._
+
+  private lazy val dataStream: InputStream = {
+    if (length == 0) {
+      new ByteArrayInputStream(new Array[Byte](0))
+    } else {
+      var is: FSDataInputStream = null
+      is = fs.open(file)
+      is.seek(offset)
+      new LimitedInputStream(is, length)
+    }
+  }
+
+  private lazy val dataInByteArray: Array[Byte] = {
+    if (length == 0) {
+      Array.empty[Byte]
+    } else {
+      var is: FSDataInputStream = null
+      try {
+        is = {
+          if (reuseFileHandle) {
+            val pathToHandleMap = handleCache.get(Thread.currentThread().getId)
+            if (pathToHandleMap == null) {
+              val res = fs.open(file)
+              handleCache.put(Thread.currentThread().getId,
+                new mutable.HashMap[Path, FSDataInputStream]() += (file -> res))
+              res
+            } else {
+              pathToHandleMap.getOrElseUpdate(file, fs.open(file))
+            }
+          } else {
+            fs.open(file)
+          }
+        }
+        is.seek(offset)
+        val array = new Array[Byte](length.toInt)
+        is.readFully(array)
+        array
+      } catch {
+        case e: IOException =>
+          var errorMessage = "Error in reading " + this
+          if (is != null) {
+            val size = fs.getFileStatus(file).getLen
+            errorMessage += " (actual file length " + size + ")"
+          }
+          throw new IOException(errorMessage, e)
+      } finally {
+        if (!reuseFileHandle) {
+          // Immediately close it if disabled file handle reuse
+          JavaUtils.closeQuietly(is)
+        }
+      }
+    }
+  }
+
+  private[spark] def prepareData(eagerRequirement: Boolean): Unit = {
+    this.eagerRequirement = eagerRequirement
+    if (! eagerRequirement) {
+      dataInByteArray
+    }
+  }
+
+  private lazy val nettyByteBuffer = ???
+
+  override def size(): Long = length
+
+  override def nioByteBuffer(): ByteBuffer = ???
```

Better to throw a meaningful exception instead of not implemented.